### PR TITLE
Introduce `!torch.tensor` / `!torch.vtensor` types.

### PR DESCRIPTION
**Pull request by silvasean**
This removes our reliance on the numpy dialect and avoids our off-label
use of the builtin tnesor type for modeling unknown dtypes.  The
`!torch.vtensor` (`ValueTensorType`) is a value-semantic tensor.  The
`!torch.tensor` (`NonValueTensorType`) and is a non-value-semantic
tensor. The new types look as follows syntactically:

```
// Least-static-information, non-value-semantic tensor.
!torch.tensor
// Explicit form of least-static-information variant.
!torch.tensor<*,unk>
// Least-static-information, Value-semantic tensor.
!torch.vtensor
// Explicit form of least-static-information variant.
!torch.vtensor<*,unk>
// Fixed-set of allowable element types, with first-class support for
// Torch's frontend signedness semantics.
!torch.tensor<*,si32>
// First-class support for unknown dtypes.
!torch.tensor<[?,?,?],unk>
// Standard MLIR representation of `?` for unknown dimensions.
!torch.tensor<[?,2,?,4],unk>
// Statically shaped / dtyped example.
!torch.vtensor<[1,2,3,4],f32>
```

This required fairly significant changes throughout the compiler, but
overall it is a big cleanup. We now have a much clearer layering of "the
Torch frontend lowering" vs "lowering to std + linalg + etc.".

At the C++ level, there is `ValueTensorType`, `NonValueTensorType`.
We also have a helper `BaseTensorType` (kind of like ShapedType) which
interoperates with those two.

Included changes:
- New `torch.tensor(dense<0.0> : tensor<5xf32>) : !torch.tensor` op for
  creating torch tensor literals in the frontend.
- Consistently use signedness for the types (except i1 which I didn't
  touch -- we need to sort out the situation with !basicpy.BoolType
  there anyway so will be attending to that soon)
- Frontend can annotate whether an argument to the function has value
  semantics. We currently require this, as our backend contract does not
  currently allow us to even model this. Before, the assumption was
  randomly injected in the middle of the pass pipeline.
- Move ArrayToTensor (now called MaximizeValueSemantics) and
  RefinePublicReturn passes to torch dialect.
- The TorchToStd and TorchToLinalg passes are now type conversions from
  `!torch.vtensor` to `tensor` and use the dialect conversion infra.
  The overall conversion pipeline is set up following the best practices
  of the "Type Conversions the Not-So-Hard Way" talk. This required
  introducing `torch-func-builtin-tensorize` and
  `torch-finalizing-builtin-tensorize` passes analogous to the upstream
  bufferization passes with the corresponding names (mostly just
  copypasta from there).
- Misc Torch-level canonicalizations -- we now cleanly layer the
  lowering to std later in the pipeline, so we are gradually lessening
  our reliance on random std constant folding before we get to that
  point.

Recommended review order:
- New types in TorchTypes.td/TorchTypes.h/TorchDialect.cpp
- New ops in TorchOps.td / TorchOps.cpp
- Less important / more mechanical stuff
  - Frontend changes.
  - Pass changes/additions in `Torch/Transforms` and `Conversion/`

### CODE REVIEW

**Comment by stellaraccident** [comment]
Diving in for a detailed review but wanted to say thank you for the detailed description. +1 on the design choices you made. I wish we'd modeled the other frontends similarly from the get-go.

File: include/npcomp/Dialect/Torch/Transforms/Passes.td [hunk 1]
```diff
@@ -164,4 +164,74 @@ def ReduceOpVariants : Pass<"torch-reduce-op-variants", "FuncOp"> {
...
+  let summary = "Refine public return";
+  let constructor = "mlir::NPCOMP::Torch::createRefinePublicReturnPass()";
+  let description = [{
+    Refines types of values returned from public functions based on
+    intraprocedural information.
+
+    This pass effectively encodes an assumption by the pass pipeline author that
+    the public calling convention of the module can have its types refined,
+    without causing ABI mismatches. This is frequently true -- for example, in
+    many systems, `tensor<?x?xf32>`, `tensor<3x3xf32>` and
```
Comment (stellaraccident) [inline include/npcomp/Dialect/Torch/Transforms/Passes.td:198]: Did you mean to be using the build-in `tensor` nomenclature here?

File: lib/Dialect/Torch/IR/TorchDialect.cpp [hunk 2]
```diff
@@ -113,5 +377,9 @@ Operation *TorchDialect::materializeConstant(OpBuilder &builder,
     if (i1Value && i1Value.getType().getIntOrFloatBitWidth() == 1)
       return builder.create<Basicpy::BoolConstantOp>(loc, type, i1Value);
   }
+  if (auto integerType = type.dyn_cast<IntegerType>()) {
+    if (integerType.getWidth() == 64)
```
Comment (stellaraccident) [inline lib/Dialect/Torch/IR/TorchDialect.cpp:381]: For my own edification: Why are you only doing this for 64 bit integers?

File: lib/Dialect/Torch/IR/TorchOps.cpp [hunk 3]
```diff
@@ -203,14 +231,40 @@ OpFoldResult Aten__Is__Op::fold(ArrayRef<Attribute> operands) {
...
-    auto buildList = op.getOperand().getDefiningOp<Basicpy::BuildListOp>();
-    if (!buildList)
-      return rewriter.notifyMatchFailure(op, "operand not basicpy.build_list");
-    rewriter.replaceOpWithNewOp<::mlir::ConstantOp>(
-        op, rewriter.getI64IntegerAttr(buildList.getNumOperands()));
+    auto size = op.getOperand().getDefiningOp<AtenSizeOp>();
+    if (!size)
+      return rewriter.notifyMatchFailure(op, "operand not AtenSizeOp");
+    // TODO: Normalize all the torch scalar integer types to consistently use
+    // one type (probably si64 or a new `!torch.int`) so that this op and others
```
Comment (stellaraccident) [inline lib/Dialect/Torch/IR/TorchOps.cpp:264]: Probably a `!torch.int`. There's already enough of a semantic gap between Python ints (which tend to be arbitrary magnitude) and Torch/et-al ints (fixed bit depth), it would be nice to preserve the fact that we are working on a `!torch.int` and not just a thing that was prematurely pinned to an `i64`

File: test/Dialect/Torch/func-builtin-tensorize.mlir [hunk 4]
```diff
@@ -0,0 +1,94 @@
+// RUN: npcomp-opt %s -torch-func-builtin-tensorize -split-input-file -verify-diagnostics -allow-unregistered-dialect | FileCheck %s
```
Comment (stellaraccident) [inline test/Dialect/Torch/func-builtin-tensorize.mlir:1]: Am I thinking about this wrong to expect a test in here with the pass operating on a !torch.tensor (even if just to verify that the wrong thing isn't done)?

**Review by stellaraccident — APPROVED** [review]
Several nits, but overall this looks quite good - and a very long time in coming overhaul to the tensor type system. Thank you.

File: frontends/pytorch/e2e_testing/torchscript/basic.py [hunk 5]
```diff
@@ -16,8 +16,8 @@ def __init__(self):
     @export
     @annotate_args([
         None,
-        ([-1, -1], torch.float32),
-        ([-1, -1], torch.float32),
+        ([-1, -1], torch.float32, True),
```
[the remark this answers is not in the document]
Comment (silvasean) [reply frontends/pytorch/e2e_testing/torchscript/basic.py:19]: Unfortunately it is a tuple, so we can't pass a kwarg. We really will need to do another pass of cleaning up this interface, and this is pushing it towards that breaking point to motivate that (thanks for keeping me honest about it though :) ). I suspect we need something like a `TensorSpec` class for these args (that we somehow digest down to the C++ layer), but that requires more dedicated attention than I could muster in the middle of this huge PR.

File: include/npcomp/Dialect/Torch/IR/TorchTypes.td [hunk 6]
```diff
@@ -44,6 +46,148 @@ def Torch_NnModuleType : Torch_Type<"NnModule", "nn.Module"> {
...
+    a strict separation between the value-semantic and potentially-mutating
+    worlds, as one of our main jobs in the compiler is to isolate the mutating
+    parts as much as possible because most lower levels of the compiler stack
+    are expected to require value semantics. E.g. npcomp's backend contract
+    is mostly in terms of linalg-on-tensor for compute-heavy ops, which require
+    a conversion to the builtin `tensor` type which has value semantics.
+    Some notes about value semantics:
+      - Using the type system described in PEP 483 (which TorchScript and other
+        Python systems follow), `!torch.tensor` is a subtype of
+        `!torch.vtensor`. Specifically, both types have the same set of values,
```
[the remark this answers is not in the document]
Comment (silvasean) [reply include/npcomp/Dialect/Torch/IR/TorchTypes.td:91]: No, it is correct as written:

- they have the same set of values (multidimensional arrays)
- !torch.tensor has additional mutating/aliasing operations

File: include/npcomp/Dialect/Torch/IR/TorchTypes.td [hunk 7]
```diff
@@ -44,6 +46,148 @@ def Torch_NnModuleType : Torch_Type<"NnModule", "nn.Module"> {
...
+    | torch.int16       | si16               |
+    | torch.int32       | si32               |
+    | torch.int64       | si64               |
+    | torch.bool        | i1                 |
+    | torch.qint8       | !torch.qint8       |
+    |-------------------|--------------------|
+    ```
+
+    TODO: Support the full set of Torch dtypes.
+    TODO: Use si1?
```
[the remark this answers is not in the document]
Comment (silvasean) [reply include/npcomp/Dialect/Torch/IR/TorchTypes.td:139]: Yeah, I think I'm going to converge on just i1 or ui1. I agree that the sign extending behavior is probably undesirable in this case.

File: include/npcomp/Dialect/Torch/Transforms/Passes.td (hunk 1 above)
Comment (silvasean) [reply include/npcomp/Dialect/Torch/Transforms/Passes.td:198]: Good catch!

File: test/Dialect/Torch/func-builtin-tensorize.mlir (hunk 4 above)
Comment (silvasean) [reply test/Dialect/Torch/func-builtin-tensorize.mlir:1]: We don't do anything specific to !torch.tensor, so it just passes through trivially (arguably we should make it an error, in which case we should add a test). In effect it would just be a redundant test of the func conversion patterns, or equivalent to a test with an i64 or other random type argument (we actually already have one like that).
